{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "lab7-bio.ipynb",
      "provenance": [],
      "collapsed_sections": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "3a. Consider the application of the Needleman-Wunsch algorithm to the protein sequences:\n",
        "S1: APSC; S2: TAPT, using the BLOSUM62 matrix and g =−4. Calculate\n",
        "the optimal alignment."
      ],
      "metadata": {
        "id": "_nVP8BR09jYO"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0RdyKFek8yl_",
        "outputId": "d6448083-6d48-421e-b830-4da5f683bd70"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Score of optimal alignment: 4\n",
            "[0, -4, -8, -12, -16]\n",
            "[-4, 0, 0, -4, -8]\n",
            "[-8, -4, -1, 7, 3]\n",
            "[-12, -7, -3, 3, 8]\n",
            "[-16, -11, -7, -1, 4]\n",
            "[0, 3, 3, 3, 3]\n",
            "[2, 1, 1, 3, 3]\n",
            "[2, 2, 1, 1, 3]\n",
            "[2, 1, 1, 2, 1]\n",
            "[2, 2, 2, 2, 2]\n",
            "-APSC\n",
            "TAPT-\n"
          ]
        }
      ],
      "source": [
        "### Implementing the Needleman-Wunsch Algorithm\n",
        "## global alignment \n",
        "def needleman_Wunsch(seq1, seq2, sm, g):\n",
        "    S = [[0]]\n",
        "    T = [[0]]\n",
        "    for j in range(1, len(seq2)+1):\n",
        "        S[0].append(g * j)\n",
        "        T[0].append(3)\n",
        "    for i in range(1, len(seq1)+1):\n",
        "        S.append([g * i])\n",
        "        T.append([2])\n",
        "    for i in range(0, len(seq1)):\n",
        "        for j in range(len(seq2)):\n",
        "            s1 = S[i][j] + score_pos (seq1[i], seq2[j], sm, g); \n",
        "            s2 = S[i][j+1] + g\n",
        "            s3 = S[i+1][j] + g\n",
        "            S[i+1].append(max(s1, s2, s3))\n",
        "            T[i+1].append(max3t(s1, s2, s3))\n",
        "            \n",
        "    return (S, T)\n",
        "\n",
        "def read_submat_file(filename):\n",
        "    sm = {}\n",
        "    f = open(filename, \"r\")\n",
        "    line = f.readline()\n",
        "    tokens = line.split(\"\\t\")\n",
        "    ns = len(tokens)\n",
        "    alphabet = []\n",
        "    for i in range(0, ns): \n",
        "        alphabet.append(tokens[i][0])\n",
        "    for i in range(0,ns):\n",
        "        line = f.readline();\n",
        "        tokens = line.split(\"\\t\");\n",
        "        for j in range(0, len(tokens)):\n",
        "            k = alphabet[i]+alphabet[j]\n",
        "            sm[k] = int(tokens[j])\n",
        "    return sm\n",
        "\n",
        "def max3t (v1, v2, v3):\n",
        "    if v1 > v2:\n",
        "        if v1 > v3: return 1\n",
        "        else: return 3\n",
        "    else:\n",
        "        if v2 > v3: return 2\n",
        "        else: return 3\n",
        "\n",
        "def recover_align (T, seq1, seq2):\n",
        "    res = [\"\", \"\"]\n",
        "    i = len(seq1)\n",
        "    j = len(seq2)\n",
        "    while i>0 or j>0:\n",
        "        if T[i][j]==1:\n",
        "            res[0] = seq1[i-1] + res[0]\n",
        "            res[1] = seq2[j-1] + res[1]\n",
        "            i -= 1\n",
        "            j -= 1\n",
        "        elif T[i][j] == 3:\n",
        "            res[0] = \"-\" + res[0]\n",
        "            res[1] = seq2[j-1] + res[1]\n",
        "            j -= 1\n",
        "        else:\n",
        "            res[0] = seq1[i-1] + res[0]\n",
        "            res[1] = \"-\" + res[1]\n",
        "            i -= 1\n",
        "    return res\n",
        "\n",
        "def print_mat (mat):\n",
        "    for i in range(0, len(mat)):\n",
        "        print(mat[i]) \n",
        "\n",
        "def score_pos (c1, c2, sm, g):\n",
        "    if c1 == \"-\" or c2==\"-\":\n",
        "        return g\n",
        "    else:\n",
        "        return sm[c1+c2]\n",
        "\n",
        "def test_global_alig():\n",
        "    sm = read_submat_file(\"./blosum62.mat\")\n",
        "    seq1 = \"APSC\"\n",
        "    seq2 = \"TAPT\"\n",
        "    res = needleman_Wunsch(seq1, seq2, sm, -4)\n",
        "    S = res[0]\n",
        "    T = res[1]\n",
        "    print(\"Score of optimal alignment:\", S[len(seq1)][len(seq2)])\n",
        "    print_mat(S)\n",
        "    print_mat(T)\n",
        "    alig = recover_align(T, seq1, seq2)\n",
        "    print(alig[0])\n",
        "    print(alig[1])\n",
        "\n",
        "test_global_alig()"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install biopython"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "sVzxwB0L-V-w",
        "outputId": "a40d85a8-7fad-4bed-b41e-88f0212efe19"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting biopython\n",
            "  Downloading biopython-1.79-cp37-cp37m-manylinux_2_5_x86_64.manylinux1_x86_64.whl (2.3 MB)\n",
            "\u001b[K     |████████████████████████████████| 2.3 MB 26.2 MB/s \n",
            "\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from biopython) (1.21.5)\n",
            "Installing collected packages: biopython\n",
            "Successfully installed biopython-1.79\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Based on the result of the previous exercise, could the following alignment be provided\n",
        "by the progressive algorithm implemented in this chapter?\n",
        "-AP-SC\n",
        "TAPT--\n",
        "TAT-S"
      ],
      "metadata": {
        "id": "kmfNZIkm-dKm"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# al_type - төрөл (ДНХ, РНХ, уураг)\n",
        "# listseqs - дарааллын жагсаалт\n",
        "# string, зайг \"-\"\n",
        "\n",
        "class MyAlign:\n",
        "\n",
        "    # init obj\n",
        "    def __init__(self, lseqs, al_type = \"protein\"):\n",
        "        self.listseqs = lseqs\n",
        "        self.al_type = al_type\n",
        "    \n",
        "    # seq 1 iin urt\n",
        "    # ooroor helbel number of columns\n",
        "    def __len__(self): \n",
        "        return len(self.listseqs[0])\n",
        "    \n",
        "    # \n",
        "    def __getitem__(self, n):\n",
        "        if type(n) is tuple and len(n) == 2: \n",
        "            i, j = n\n",
        "            return self.listseqs[i][j]\n",
        "        elif type(n) is int: return self.listseqs[n]\n",
        "        return None\n",
        "    \n",
        "    # listiig haritsuulj harah\n",
        "    def __str__(self):\n",
        "        res = \"\"\n",
        "        for seq in self.listseqs:\n",
        "            res += \"\\n\" + seq \n",
        "        return res\n",
        "    \n",
        "    # listiin urtiig \n",
        "    # matrix = 2\n",
        "    def num_seqs(self):\n",
        "        return len(self.listseqs)\n",
        "   \n",
        "    # \n",
        "    def column (self, indice):\n",
        "        res = []\n",
        "        for k in range(len(self.listseqs)):\n",
        "            res.append(self.listseqs[k][indice])\n",
        "        return res\n",
        "    \n",
        "    # Бас нэг чухал метод бол Зэрэгцүүлэлтийн тогтворжон хэлбэрийг үүсгэх тооцоолол\n",
        "    # Тогтворжсон хэлбэрийг зэрэгцүүлэлтийн багана бүрийн хувьд зайг тооцоогүй хамгийн өндөр\n",
        "    # давтамжтай тэмдэгтүүдийн дараалалаар илэрхийлнэ.\n",
        "    # Энэ метод нь dictionary ашиглан багана бүрийн тэмдэгтүүдийн давтамжийг тоолж, хамгийн олныг сонгоно\n",
        "    def consensus (self):\n",
        "        cons = \"\"\n",
        "        for i in range(len(self)):\n",
        "            cont = {}\n",
        "            for k in range(len(self.listseqs)):\n",
        "                c = self.listseqs[k][i]\n",
        "                if c in cont:\n",
        "                    cont[c] = cont[c] + 1\n",
        "                else: \n",
        "                    cont[c] = 1\n",
        "            maximum = 0\n",
        "            cmax = None\n",
        "            for ke in cont.keys():\n",
        "                if ke != \"-\" and cont[ke] > maximum: \n",
        "                    maximum = cont[ke]\n",
        "                    cmax = ke\n",
        "            cons = cons + cmax\n",
        "        return cons\n",
        "alig = MyAlign([\"ATGA-A\",\"AA-AT-\"],\"dna\")\n",
        "print(alig.consensus())"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "4KWaaM53BNyr",
        "outputId": "99dba57a-7601-4536-953d-338cbde2a8c5"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "ATGATA\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Dictionary бүтцэд суурилан дүрсэлсэн орлуулах матрицыг ашиглана.\n",
        "# цагаан толгойг хадгалдаг аттрибут\n",
        "# хос тэмдэгтүүдийн оноотой dictionary\n",
        "# Өмнө нь үзсэн кодтой ижил төстэй гүйцээж бичнэч\n",
        "class SubstMatrix:\n",
        "\n",
        "    def __init__(self):\n",
        "        self.alphabet = \"\"\n",
        "        self.sm = {}\n",
        "        \n",
        "    def __getitem__(self, ij):\n",
        "        i, j = ij\n",
        "        return self.score_pair(i, j)\n",
        "    \n",
        "    def score_pair(self, c1, c2):\n",
        "        if c1 not in self.alphabet or c2 not in self.alphabet:\n",
        "            return None\n",
        "        return self.sm[c1+c2]\n",
        "        \n",
        "    def read_submat_file(self, filename, sep):\n",
        "        f = open(filename, \"r\")\n",
        "        line = f.readline()\n",
        "        tokens = line.split(sep)\n",
        "        ns = len(tokens)\n",
        "        self.alphabet = \"\"\n",
        "        for i in range(0, ns): \n",
        "            self.alphabet += tokens[i][0]\n",
        "        for i in range(0,ns):\n",
        "            line = f.readline();\n",
        "            tokens = line.split(sep);\n",
        "            for j in range(0, len(tokens)):\n",
        "                k = self.alphabet[i]+self.alphabet[j]\n",
        "                self.sm[k] = int(tokens[j])\n",
        "        f.close()\n",
        "        return None\n",
        "                \n",
        "    def create_submat(self, match, mismatch, alphabet):\n",
        "        self.alphabet = alphabet\n",
        "        for c1 in alphabet:\n",
        "            for c2 in alphabet:\n",
        "                if (c1 == c2):\n",
        "                    self.sm[c1+c2] = match\n",
        "                else:\n",
        "                    self.sm[c1+c2] = mismatch\n",
        "        return None"
      ],
      "metadata": {
        "id": "jvb7dod2E8Bi"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# орлуулах матриц ба зай-торгууль\n",
        "# зэрэгцүүлэх дарааллууд,\n",
        "# DP алгоритмын S ба T матрицууд.\n",
        "class PairwiseAlignment:\n",
        "\n",
        "    def __init__(self, sm, g):\n",
        "        self.g = g\n",
        "        self.sm = sm\n",
        "        self.S = None\n",
        "        self.T = None\n",
        "        self.seq1 = None\n",
        "        self.seq2 = None\n",
        "        \n",
        "    def score_pos (self, c1, c2):\n",
        "        if c1 == \"-\" or c2==\"-\":\n",
        "            return self.g\n",
        "        else:\n",
        "            return self.sm[c1,c2]\n",
        "        \n",
        "    def score_alin (self, alin):\n",
        "        res = 0;\n",
        "        for i in range(len(alin)):\n",
        "            res += self.scorePos (alin[0][i], alin[1][i])\n",
        "        return res\n",
        "    \n",
        "    def needleman_Wunsch (self, seq1, seq2):\n",
        "        if (seq1.seq_type != seq2.seq_type): return None\n",
        "        self.S = [[0]]\n",
        "        self.T = [[0]]\n",
        "        self.seq1 = seq1\n",
        "        self.seq2 = seq2\n",
        "        for j in range(1, len(seq2)+1):\n",
        "            self.S[0].append(self.g * j)\n",
        "            self.T[0].append(3)\n",
        "        for i in range(1, len(seq1)+1):\n",
        "            self.S.append([self.g * i])\n",
        "            self.T.append([2])\n",
        "        for i in range(0, len(seq1)):\n",
        "            for j in range(len(seq2)):\n",
        "                s1 = self.S[i][j] + self.score_pos(seq1[i], seq2[j])\n",
        "                s2 = self.S[i][j+1] + self.g\n",
        "                s3 = self.S[i+1][j] + self.g\n",
        "                self.S[i+1].append(max(s1, s2, s3))\n",
        "                self.T[i+1].append(max3t(s1, s2, s3))\n",
        "        return self.S[len(seq1)][len(seq2)]\n",
        "    \n",
        "    def recover_align (self):\n",
        "        res = [\"\", \"\"]\n",
        "        i = len(self.seq1)\n",
        "        j = len(self.seq2)\n",
        "        while i>0 or j>0:\n",
        "            if self.T[i][j]==1:\n",
        "                res[0] = self.seq1[i-1] + res[0]\n",
        "                res[1] = self.seq2[j-1] + res[1]\n",
        "                i -= 1\n",
        "                j -= 1\n",
        "            elif self.T[i][j] == 3:\n",
        "                res[0] = \"-\" + res[0]\n",
        "                res[1] = self.seq2[j-1] + res[1] \n",
        "                j -= 1\n",
        "            else:\n",
        "                res[0] = self.seq1[i-1] + res[0]\n",
        "                res[1] = \"-\" + res[1]\n",
        "                i -= 1\n",
        "        return MyAlign(res, self.seq1.seq_type)\n",
        "     \n",
        "    def smith_Waterman (self, seq1, seq2):\n",
        "        if (seq1.seq_type != seq2.seq_type): return None\n",
        "        self.S = [[0]]\n",
        "        self.T = [[0]]\n",
        "        self.seq1 = seq1\n",
        "        self.seq2 = seq2\n",
        "        maxscore = 0\n",
        "        for j in range(1, len(seq2)+1):\n",
        "            self.S[0].append(0)\n",
        "            self.T[0].append(0)\n",
        "        for i in range(1, len(seq1)+1):\n",
        "            self.S.append([0])\n",
        "            self.T.append([0])\n",
        "        for i in range(0, len(seq1)):\n",
        "            for j in range(len(seq2)):\n",
        "                s1 = self.S[i][j] + self.score_pos(seq1[i], seq2[j]) \n",
        "                s2 = self.S[i][j+1] + self.g\n",
        "                s3 = self.S[i+1][j] + self.g\n",
        "                b = max(s1, s2, s3)\n",
        "                if b <= 0:\n",
        "                    self.S[i+1].append(0)\n",
        "                    self.T[i+1].append(0)\n",
        "                else:\n",
        "                    self.S[i+1].append(b)\n",
        "                    self.T[i+1].append(max3t(s1, s2, s3))\n",
        "                    if b > maxscore: \n",
        "                        maxscore = b\n",
        "        return maxscore\n",
        "\n",
        "    def recover_align_local (self):\n",
        "        res = [\"\", \"\"]\n",
        "        maxscore = 0\n",
        "        maxrow = 0\n",
        "        maxcol = 0\n",
        "        for i in range(1,len(self.S)):\n",
        "            for j in range(1, len(self.S[i])):\n",
        "                if self.S[i][j] > maxscore:\n",
        "                    maxscore = self.S[i][j]\n",
        "                    maxrow = i\n",
        "                    maxcol = j\n",
        "        i = maxrow\n",
        "        j = maxcol\n",
        "        while i>0 or j>0:\n",
        "            if self.T[i][j]==1:\n",
        "                res[0] = self.seq1[i-1] + res[0]\n",
        "                res[1] = self.seq2[j-1] + res[1]\n",
        "                i -= 1\n",
        "                j -= 1\n",
        "            elif self.T[i][j] == 3:\n",
        "                res[0] = \"-\" + res[0];\n",
        "                res[1] = self.seq2[j-1] + res[1]; \n",
        "                j -= 1\n",
        "            elif self.T[i][j] == 2:\n",
        "                res[0] = self.seq1[i-1] + res[0];\n",
        "                res[1] = \"-\" + res[1]; \n",
        "                i -= 1\n",
        "            else: break\n",
        "        return MyAlign(res, self.seq1.seq_type)\n",
        "\n",
        "\n",
        "def max3t (v1, v2, v3):\n",
        "    if v1 > v2:\n",
        "        if v1 > v3: return 1\n",
        "        else: return 3\n",
        "    else:\n",
        "        if v2 > v3: return 2\n",
        "        else: return 3\n",
        "\n",
        "def printMat (mat):\n",
        "    for i in range(0, len(mat)):\n",
        "        print(mat[i])"
      ],
      "metadata": {
        "id": "gM0uSn1OFUp5"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# -*- coding: utf-8 -*-\n",
        "\n",
        "def translate_codon (cod):\n",
        "    \"\"\"Translates a codon into an aminoacid using an internal dictionary with the standard genetic code.\"\"\"\n",
        "    tc = {\"GCT\":\"A\", \"GCC\":\"A\", \"GCA\":\"A\", \"GCG\":\"A\", \n",
        "      \"TGT\":\"C\", \"TGC\":\"C\",\n",
        "      \"GAT\":\"D\", \"GAC\":\"D\",\n",
        "      \"GAA\":\"E\", \"GAG\":\"E\",\n",
        "      \"TTT\":\"F\", \"TTC\":\"F\",\n",
        "      \"GGT\":\"G\", \"GGC\":\"G\", \"GGA\":\"G\", \"GGG\":\"G\",\n",
        "      \"CAT\":\"H\", \"CAC\":\"H\",\n",
        "      \"ATA\":\"I\", \"ATT\":\"I\", \"ATC\":\"I\",\n",
        "      \"AAA\":\"K\", \"AAG\":\"K\",\n",
        "      \"TTA\":\"L\", \"TTG\":\"L\", \"CTT\":\"L\", \"CTC\":\"L\", \"CTA\":\"L\", \"CTG\":\"L\",\n",
        "      \"ATG\":\"M\", \"AAT\":\"N\", \"AAC\":\"N\",\n",
        "      \"CCT\":\"P\", \"CCC\":\"P\", \"CCA\":\"P\", \"CCG\":\"P\",\n",
        "      \"CAA\":\"Q\", \"CAG\":\"Q\",\n",
        "      \"CGT\":\"R\", \"CGC\":\"R\", \"CGA\":\"R\", \"CGG\":\"R\", \"AGA\":\"R\", \"AGG\":\"R\",\n",
        "      \"TCT\":\"S\", \"TCC\":\"S\", \"TCA\":\"S\", \"TCG\":\"S\", \"AGT\":\"S\", \"AGC\":\"S\",\n",
        "      \"ACT\":\"T\", \"ACC\":\"T\", \"ACA\":\"T\", \"ACG\":\"T\",\n",
        "      \"GTT\":\"V\", \"GTC\":\"V\", \"GTA\":\"V\", \"GTG\":\"V\",\n",
        "      \"TGG\":\"W\",\n",
        "      \"TAT\":\"Y\", \"TAC\":\"Y\",\n",
        "      \"TAA\":\"_\", \"TAG\":\"_\", \"TGA\":\"_\"}\n",
        "    if cod in tc: return tc[cod]\n",
        "    else: return None\n",
        "\n",
        "class MySeq: \n",
        "    \"\"\" Class for biological sequences. \"\"\"\n",
        "    \n",
        "    def __init__ (self, seq, seq_type = \"DNA\"): \n",
        "        self.seq = seq.upper()\n",
        "        self.seq_type = seq_type\n",
        "\n",
        "    def __len__(self):\n",
        "        return len(self.seq)\n",
        "    \n",
        "    def __getitem__(self, n):\n",
        "        return self.seq[n]\n",
        "\n",
        "    def __getslice__(self, i, j):\n",
        "        return self.seq[i:j]\n",
        "\n",
        "    def __str__(self):\n",
        "        return self.seq\n",
        "        \n",
        "    def get_seq_biotype (self):\n",
        "        return self.seq_type\n",
        "        \n",
        "    def show_info_seq (self):\n",
        "        print (\"Sequence: \" + self.seq + \" biotype: \" + self.seq_type)\n",
        "        \n",
        "    def alphabet (self):\n",
        "        if (self.seq_type==\"DNA\"): return \"ACGT\"\n",
        "        elif (self.seq_type==\"RNA\"): return \"ACGU\"\n",
        "        elif (self.seq_type==\"PROTEIN\"): return \"ACDEFGHIKLMNPQRSTVWY\"\n",
        "        else: return None\n",
        "        \n",
        "    def validate (self):\n",
        "        alp = self.alphabet()\n",
        "        res = True\n",
        "        i = 0\n",
        "        while i < len(self.seq) and res:\n",
        "            if self.seq[i] not in alp: res = False\n",
        "            else: i += 1\n",
        "        return res \n",
        "        \n",
        "    def transcription (self):\n",
        "        if (self.seq_type == \"DNA\"):\n",
        "            return MySeq(self.seq.replace(\"T\",\"U\"), \"RNA\")\n",
        "        else:\n",
        "            return None\n",
        "    \n",
        "    def reverse_comp (self):\n",
        "        if (self.seq_type != \"DNA\"): return None\n",
        "        comp = \"\"\n",
        "        for c in self.seq:\n",
        "            if (c == 'A'): comp = \"T\" + comp \n",
        "            elif (c == \"T\"): comp = \"A\" + comp \n",
        "            elif (c == \"G\"): comp = \"C\" + comp\n",
        "            elif (c== \"C\"): comp = \"G\" + comp\n",
        "        return MySeq(comp, \"DNA\")\n",
        "        \n",
        "    def translate (self, iniPos= 0):\n",
        "        if (self.seq_type != \"DNA\"): return None\n",
        "        seq_aa = \"\"\n",
        "        for pos in range(iniPos,len(self.seq)-2,3):\n",
        "            cod = self.seq[pos:pos+3]\n",
        "            seq_aa += translate_codon(cod)\n",
        "        return MySeq(seq_aa, \"PROTEIN\")      "
      ],
      "metadata": {
        "id": "sIQ1z9PfGZlr"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# seqs – зэрэгцүүлсэн дарааллууд\n",
        "# alignpars – Needleman-Wunsch алгоритмд хэрэглэгдэх зэрэгцүүлэлтийн параметерүүд\n",
        "\n",
        "class MultipleAlignment():\n",
        "\n",
        "    def __init__(self, seqs, alignseq):\n",
        "        self.seqs = seqs\n",
        "        self.alignpars = alignseq\n",
        "    \n",
        "    def num_seqs(self):\n",
        "        return len(self.seqs)\n",
        "    \n",
        "    def add_seq_alignment (self, alignment, seq):\n",
        "        res = []\n",
        "        for i in range(len(alignment.listseqs)+1):\n",
        "            res.append(\"\")\n",
        "        cons = MySeq(alignment.consensus(),alignment.al_type)\n",
        "        self.alignpars.needleman_Wunsch(cons, seq)\n",
        "        align2 = self.alignpars.recover_align()\n",
        "        orig = 0\n",
        "        for i in range(len(align2)):\n",
        "            if align2[0,i]== '-':\n",
        "                for k in range(len(alignment.listseqs)):\n",
        "                    res[k] += \"-\"\n",
        "            else:\n",
        "                for k in range(len(alignment.listseqs)):\n",
        "                    res[k] += alignment[k,orig]\n",
        "                orig+=1\n",
        "        res[len(alignment.listseqs)] = align2.listseqs[1]\n",
        "        return MyAlign(res, alignment.al_type)\n",
        "    \n",
        "    def align_consensus(self):\n",
        "        self.alignpars.needleman_Wunsch(self.seqs[0], self.seqs[1])\n",
        "        res = self.alignpars.recover_align()\n",
        "\n",
        "        for i in range(2, len(self.seqs)):\n",
        "            res = self.add_seq_alignment(res, self.seqs[i])\n",
        "        return res\n",
        "    \n",
        "\n",
        "def printMat (mat):\n",
        "    for i in range(0, len(mat)):\n",
        "        print(mat[i])\n"
      ],
      "metadata": {
        "id": "lrZsiUSUFV1W"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "s1 = MySeq(\"ACATATCAT\")\n",
        "s2 = MySeq(\"AACAGATCT\")\n",
        "s3 = MySeq(\"AGATATTAG\")\n",
        "s4 = MySeq(\"GCATCGATT\")\n",
        "\n",
        "sm = SubstMatrix()\n",
        "sm.create_submat(1,-1,\"ACGT\") # match, mismatch, aplhabet\n",
        "aseq = PairwiseAlignment(sm,-1)# gap\n",
        "ma = MultipleAlignment([s1,s2,s3,s4], aseq)\n",
        "print(ma.align_consensus())"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "fPOdbyNWGOGJ",
        "outputId": "a9d43d49-cbb4-4ffb-e151-d68a38036705"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "A-CATATC-AT-\n",
            "AACAGATC--T-\n",
            "AG-ATATT-AG-\n",
            "G-CAT--CGATT\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio import Align\n",
        "aligner = Align.PairwiseAligner()\n",
        "alignments = aligner.align(\"APSC\", \"TAPT\")\n",
        "alignment = alignments[0]\n",
        "print(alignment)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2lQ-dGVfJetw",
        "outputId": "6d3daddd-09eb-4706-95fc-eb097115aee1"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "-APSC-\n",
            "-||---\n",
            "TAP--T\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio.Seq import Seq\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "from Bio.Align import MultipleSeqAlignment\n",
        "a = SeqRecord(Seq(\"APSC\"), id=\"Alpha\")\n",
        "b = SeqRecord(Seq(\"TAPT\"), id=\"Beta\")\n",
        "align = MultipleSeqAlignment([a, b])\n",
        "print(align)\n",
        "print (align[1])"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "BbtSuzGx-g0o",
        "outputId": "daf14e06-ae05-4fa4-b769-c9760d3ea3f2"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Alignment with 2 rows and 4 columns\n",
            "APSC Alpha\n",
            "TAPT Beta\n",
            "ID: Beta\n",
            "Name: <unknown name>\n",
            "Description: <unknown description>\n",
            "Number of features: 0\n",
            "Seq('TAPT')\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Write a method to add to the class MyAlign that, given an alignment (self ), returns the\n",
        "list of columns (indexes) in the alignment that are rich in polar basic aminoacids (R, H,\n",
        "or K). To be considered rich, the column needs to include at least half of the aminoacids\n",
        "in this group."
      ],
      "metadata": {
        "id": "7ZRF9QY9OV-V"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# al_type - төрөл (ДНХ, РНХ, уураг)\n",
        "# listseqs - дарааллын жагсаалт\n",
        "# string, зайг \"-\"\n",
        "\n",
        "class MyAlign:\n",
        "\n",
        "    # init obj\n",
        "    def __init__(self, lseqs, al_type = \"protein\"):\n",
        "        self.listseqs = lseqs\n",
        "        self.al_type = al_type\n",
        "    \n",
        "    # seq 1 iin urt\n",
        "    # ooroor helbel number of columns\n",
        "    def __len__(self): \n",
        "        return len(self.listseqs[0])\n",
        "    \n",
        "    def __getitem__(self, n):\n",
        "        if type(n) is tuple and len(n) == 2: \n",
        "            i, j = n\n",
        "            return self.listseqs[i][j]\n",
        "        elif type(n) is int: return self.listseqs[n]\n",
        "        return None\n",
        "    \n",
        "    # listiig haritsuulj harah\n",
        "    def __str__(self):\n",
        "        res = \"\"\n",
        "        for seq in self.listseqs:\n",
        "            res += \"\\n\" + seq \n",
        "        return res\n",
        "    \n",
        "    # listiin urtiig \n",
        "    # matrix = 2\n",
        "    def num_seqs(self):\n",
        "        return len(self.listseqs)\n",
        "   \n",
        "    def column (self, indice):\n",
        "        res = []\n",
        "        for k in range(len(self.listseqs)):\n",
        "            res.append(self.listseqs[k][indice])\n",
        "        return res\n",
        "\n",
        "    def rich_amino(self):\n",
        "      count = 0\n",
        "      cols = []\n",
        "      l = min(self.listseqs, key=len)\n",
        "      for k in range(len(self.listseqs)):\n",
        "        for j in range(len(l)):\n",
        "          res = self.listseqs[k][j]\n",
        "          if(\"R\" in res or \"H\" in res or \"K\" in res):\n",
        "            count += 1\n",
        "            if(count >= (len(l))/2):\n",
        "              cols.append(j)\n",
        "      return cols\n",
        "    \n",
        "    # Бас нэг чухал метод бол Зэрэгцүүлэлтийн тогтворжон хэлбэрийг үүсгэх тооцоолол\n",
        "    # Тогтворжсон хэлбэрийг зэрэгцүүлэлтийн багана бүрийн хувьд зайг тооцоогүй хамгийн өндөр\n",
        "    # давтамжтай тэмдэгтүүдийн дараалалаар илэрхийлнэ.\n",
        "    # Энэ метод нь dictionary ашиглан багана бүрийн тэмдэгтүүдийн давтамжийг тоолж, хамгийн олныг сонгоно\n",
        "    def consensus (self):\n",
        "        cons = \"\"\n",
        "        for i in range(len(self)):\n",
        "            cont = {}\n",
        "            for k in range(len(self.listseqs)):\n",
        "                c = self.listseqs[k][i]\n",
        "                if c in cont:\n",
        "                    cont[c] = cont[c] + 1\n",
        "                else: \n",
        "                    cont[c] = 1\n",
        "            maximum = 0\n",
        "            cmax = None\n",
        "            for ke in cont.keys():\n",
        "                if ke != \"-\" and cont[ke] > maximum: \n",
        "                    maximum = cont[ke]\n",
        "                    cmax = ke\n",
        "            cons = cons + cmax\n",
        "        return cons\n",
        "\n",
        "alig = MyAlign([\"ATRK\",\"ARRR\"],\"dna\")\n",
        "print(alig.column(2))\n",
        "print(alig.consensus())"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "PiDnvCvvOWLx",
        "outputId": "bb5bfc27-364b-4c17-b1f4-c0f3af20efb4"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "['R', 'R']\n",
            "ATRK\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "print(alig.rich_amino())"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Geo-D_vWSu8W",
        "outputId": "92dfa6ed-75f6-44ae-ea2b-8f11b28b8e52"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[3, 1, 2, 3]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Write a method to add to the class MyAlign that, given an alignment (self ), returns a\n",
        "string with a symbol for each column of the alignment, following these rules: ‘*’, if the\n",
        "column is fully conserved (it has all symbols equal and no gaps); ‘:’, if the column has at\n",
        "least half of the symbols equal; ‘.’: if the column does not match any of the previous, but\n",
        "has no gaps; ‘ ’, in all other cases."
      ],
      "metadata": {
        "id": "Xo0i0viEVMGe"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# al_type - төрөл (ДНХ, РНХ, уураг)\n",
        "# listseqs - дарааллын жагсаалт\n",
        "# string, зайг \"-\"\n",
        "\n",
        "class MyAlign:\n",
        "\n",
        "    # init obj\n",
        "    def __init__(self, lseqs, al_type = \"protein\"):\n",
        "        self.listseqs = lseqs\n",
        "        self.al_type = al_type\n",
        "    \n",
        "    # seq 1 iin urt\n",
        "    # ooroor helbel number of columns\n",
        "    def __len__(self): \n",
        "        return len(self.listseqs[0])\n",
        "    \n",
        "    # \n",
        "    def __getitem__(self, n):\n",
        "        if type(n) is tuple and len(n) == 2: \n",
        "            i, j = n\n",
        "            return self.listseqs[i][j]\n",
        "        elif type(n) is int: return self.listseqs[n]\n",
        "        return None\n",
        "    \n",
        "    # listiig haritsuulj harah\n",
        "    def __str__(self):\n",
        "        res = \"\"\n",
        "        for seq in self.listseqs:\n",
        "            res += \"\\n\" + seq \n",
        "        return res\n",
        "    \n",
        "    # listiin urtiig \n",
        "    # matrix = 2\n",
        "    def num_seqs(self):\n",
        "        return len(self.listseqs)\n",
        "   \n",
        "    # \n",
        "    def column (self, indice):\n",
        "        res = []\n",
        "        for k in range(len(self.listseqs)):\n",
        "            res.append(self.listseqs[k][indice])\n",
        "        return res\n",
        "    \n",
        "    # Бас нэг чухал метод бол Зэрэгцүүлэлтийн тогтворжон хэлбэрийг үүсгэх тооцоолол\n",
        "    # Тогтворжсон хэлбэрийг зэрэгцүүлэлтийн багана бүрийн хувьд зайг тооцоогүй хамгийн өндөр\n",
        "    # давтамжтай тэмдэгтүүдийн дараалалаар илэрхийлнэ.\n",
        "    # Энэ метод нь dictionary ашиглан багана бүрийн тэмдэгтүүдийн давтамжийг тоолж, хамгийн олныг сонгоно\n",
        "    def consensus (self):\n",
        "        cons = \"\"\n",
        "        for i in range(len(self)):\n",
        "            cont = {}\n",
        "            for k in range(len(self.listseqs)):\n",
        "                c = self.listseqs[k][i]\n",
        "                if c in cont:\n",
        "                    cont[c] = cont[c] + 1\n",
        "                else: \n",
        "                    cont[c] = 1\n",
        "            maximum = 0\n",
        "            cmax = None\n",
        "            for ke in cont.keys():\n",
        "                if ke != \"-\" and cont[ke] > maximum: \n",
        "                    maximum = cont[ke]\n",
        "                    cmax = ke\n",
        "            cons = cons + cmax\n",
        "        return cons\n",
        "alig = MyAlign([\"ATGA-A\",\"AA-AT-\"],\"dna\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "_SRagcCeVaxc",
        "outputId": "446fe0c5-f456-4a5a-ee51-1fb88c2bc598"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "0\n",
            "A\n",
            "A\n",
            "1\n",
            "T\n",
            "A\n",
            "2\n",
            "G\n",
            "-\n",
            "3\n",
            "A\n",
            "A\n",
            "4\n",
            "-\n",
            "T\n",
            "5\n",
            "A\n",
            "-\n",
            "None\n"
          ]
        }
      ]
    }
  ]
}